sphinx/builders/__init__.py

```diff
@@ -485,6 +485,7 @@ def read_doc(self, docname: str, *, _cache: bool = True) -> None:
         filename = self.env.doc2path(docname)
         filetype = get_filetype(self.app.config.source_suffix, filename)
         publisher = self.app.registry.get_publisher(self.app, filetype)
+        self.env.temp_data['parser'] = publisher.parser
```

Comment:
This one might conflict an existing `temp_data` specified by the user somewhere else (unless `temp_data` is only meant for us). In particular, this requires a CHANGES entry.

I think we should improve the way we store temporary data and use namespaces to avoid conflicts.

Reply:
I changed the key to `_parser`, which should minimise the risk of conflict, and exposed it as a property of the `BuildEnvironment`, similar to `docname`

Also added a CHANGES entry

Reply:
> I think we should improve the way we store temporary data and use namespaces to avoid conflicts.

well thats a topic for another day 😅